feat: slashing release

stevennevins opened the pull request.

gpsanant committed:
Update LICENSE (#285)

stevennevins committed:
chore: upgrade core to target operator set release

stevennevins committed:
Merge branch 'dev' into feat/operate-set-release-branch

shrimalmadhur committed:
chore: update mock contracts with latest interfaces (#293)

* chore: update mock contracts with latest interfaces

stevennevins committed:
chore: bump to latest operator set release commits

stevennevins committed:
chore: fixes for depedency bump

stevennevins committed:
feat: operator set migration-1-migration (#286)

* chore: checkout migration branch

* feat: implement migration function

* chore: update to release branch

* feat: operator set creation for each quorum number

* feat: migration with merge sorted array of operators and their quorums

* feat: operator set migration working

* chore: revert change from testing

* chore: revert change from testing

* chore: remove extra logging and commented out asserts

* chore: remove unused file

* fix: remove console logs

* refactor: to view functions

* chore: nit and remove unneeded function

* fix: remove duplication of looping for all the operators

* chore: remove comment

* feat: allow migrating in two transactions

* feat: finalization of migration

* chore: use string errors and fix migration issues

* chore: rename

* feat: use library for merge sort

* test: fuzz view function

stevennevins committed:
feat: operator set migration-2-create quorum (#287)

* chore: checkout migration branch

* feat: implement migration function

* chore: update to release branch

* feat: operator set creation for each quorum number

* feat: migration with merge sorted array of operators and their quorums

* feat: operator set migration working

* chore: revert change from testing

* chore: revert change from testing

* chore: remove extra logging and commented out asserts

* chore: remove unused file

* fix: remove console logs

* feat: create quorum post operator set migration

* test(wip): create quorum test adds new operator set

* test: migration create quorum

* refactor: to view functions

* chore: nit and remove unneeded function

* fix: remove duplication of looping for all the operators

* chore: remove comment

* feat: allow migrating in two transactions

* feat: finalization of migration

* chore: use string errors and fix migration issues

* chore: rename

* feat: use library for merge sort

* test: fuzz view function

* fix: updates from merge

* chore: use interface

stevennevins committed:
chore: bump operator set release dependency in core

stevennevins committed:
chore: updates from dependency bump

stevennevins committed:
feat: add natspec

stevennevins committed:
docs: add natspec

stevennevins committed:
feat: add checks operator was registered for quorums when migrating

stevennevins committed:
fix: resolve code size issue in RegistryCoordinator

shrimalmadhur committed:
chore: update to latest core (#299)

* chore: update to latest eigenlayer-contracts feat/operator-set-release

* chore: add method to mock

stevennevins committed:
feat(op sets): update stakes when forceUnregister (#300)

* feat: update stakes handle direct deregistration on AVSDirectory

* test: update stake for quorum if operator directly unregistered from the AVSDirectory

* chore: simplify setup

* chore: simplify setup

* chore: make service manager immutable on stakeRegistry

stevennevins committed:
merge dev

stevennevins committed:
feat(op sets): register and deregister (#301)

* feat: register and deregister to operator sets

* fix: bytecode wrangling

* chore: shorter errors to reduce bytecode

* test: register after migration

* chore: remove stale todos

* chore: add back commented line

* chore: remain consistent with m2 events

* fix: side effect of merge from _deregister function

* fix: bytecode massaging

* chore: rename for clarity

* chore: remove comments and whitespace

* docs: add natspec to the library

stevennevins committed:
feat(op sets): upgrade and migrate script (#303)

* feat: upgrade and test pre prod upgrade and migration

* fix: add param introduced in merge

stevennevins committed:
chore: bump slashing core dependency (#312)

* chore: bump to slashing branch

* chore: bump compiler version

* fix: dep interface changes

* fix: compiler errors from interface changes and type changes

* fix: compiler errors

* chore: bump dependencies

* chore: bump core dependency and resolve issues

* chore: bump core dependency and fix compiler errors

* feat: integrate AllocationManager

* feat: add a slashing permission to the service manager

* chore: remove unneeded casting

* feat: implement a slasher permission and forward call to AllocationManager

* feat: add simiple slasher starting point

* chore: bump slashing magnitudes

* chore: bump core slashing-magnitudes branch

stevennevins committed:
feat: slasher templates / examples (#310)

* chore: bump to slashing branch

* chore: bump compiler version

* fix: dep interface changes

* fix: compiler errors from interface changes and type changes

* fix: compiler errors

* chore: bump dependencies

* chore: bump core dependency and resolve issues

* chore: bump core dependency and fix compiler errors

* feat: integrate AllocationManager

* feat: add a slashing permission to the service manager

* chore: remove unneeded casting

* feat: implement a slasher permission and forward call to AllocationManager

* feat: add simiple slasher starting point

* feat: slashers

* chore: change around slashed event

* fix: call dm

* feat: add proposal mechanism for updating slasher

* fix: set to completed instead of delete

* chore: use struct instead of params directly

* chore: clean up params more

* chore: simplify and organize files

* chore: cleanup logic and couple event with internal func

* fix: pass correct params

* chore: organize and add interface

* chore: nits

* chore: cleanup more nits

* fix: storage gap

* chore: nits refactor

* chore: go back to fulfill being onlySlasher

* test: fixes from core updates

* fix: use delegated stake per operator set instead of per AVS

* fix: update to 14 days

* feat: configurable lookahead and stake type

wadealexc commented on src/ServiceManagerBaseStorage.sol:
```diff
@@ -35,19 +37,33 @@ abstract contract ServiceManagerBaseStorage is IServiceManager, OwnableUpgradeab
 
     // storage gap for upgradeability
-    uint256[49] private __GAP;
+    uint256[46] private __GAP;
```
Looks like GAP should be decreased by one additional?

stevennevins replied:
yup nice catch

stevennevins replied:
[fa04f06](https://github.com/Layr-Labs/eigenlayer-middleware/pull/311/commits/fa04f06212978dc3c66c23976cd34498f73aad41)

wadealexc commented on src/StakeRegistry.sol:
```diff
@@ -471,14 +504,17 @@ contract StakeRegistry is StakeRegistryStorage {
+        address[] memory operators = new address[](1);
+        operators[0] = operator;
+        uint32 beforeTimestamp = uint32(block.timestamp + slashableStakeLookAhead);
+        (uint256[][] memory strategyShares, ) = IAllocationManager(serviceManager.allocationManager()).getMinDelegatedAndSlashableOperatorShares(OperatorSet(address(serviceManager), quorumNumber), operators ,strategiesPerQuorum[quorumNumber], beforeTimestamp);
```
Was removed in core

stevennevins committed:
feat: remove both option

stevennevins committed:
chore: remove unused test util contracts (#319)

* feat: remove both option

* chore: remove unused test util contracts

* chore: remove diff

stevennevins committed:
fix: storage gap remove one slot (#320)

stevennevins committed:
feat: track total slashable stake and total delegated stake per quorum (#317)

* feat: remove both option

* feat: total delegated stake and total slashable stake per quorum config

* test: resolve some breaking changes to tests

* chore: move stake type to file level definition

* chore: refactor loop

* test: add unit test for slashble stake quorum init

* test: assert on state and event

* test: delegated stake quorum and assertions

supernovahs commented:
```
Error (9553): Invalid type for argument in function call. Invalid implicit conversion from contract TransparentUpgradeableProxy to contract ITransparentUpgradeableProxy requested.
   --> lib/eigenlayer-middleware/test/utils/MockAVSDeployer.sol:310:17:
    |
310 |                 TransparentUpgradeableProxy(payable(address(registryCoordinator))),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Error: Compilation failed
```
getitng this while compiling my contracts in IS with this as submodule

supernovahs commented on test/mocks/ECDSAServiceManagerMock.sol:
```diff
@@ -19,4 +27,18 @@ contract ECDSAServiceManagerMock is ECDSAServiceManagerBase {
+
+    function createOperatorSets(uint32[] memory) external {}
+
+    function registerOperatorToOperatorSets(
+        address operator,
+        uint32[] calldata operatorSetIds,
+        ISignatureUtils.SignatureWithSaltAndExpiry memory operatorSignature
+    ) external {}
+
+    function deregisterOperatorFromOperatorSets(address operator, uint32[] calldata operatorSetIds) external{}
```
these can be directly called by the operator right? Or is it necessary to route these through service manager?

stevennevins replied:
These have to be routed through the service manager currently, but that will likely change post-UAM  

They won't be called directly by the operator on the service manager, though.  The flow would be

Registry -> ServiceManager -> AllocationManager

supernovahs replied:
what is UAM?

stevennevins replied:
universal access managemnent

supernovahs reviewed (commented):
^

stevennevins commented:
> ```
> Error (9553): Invalid type for argument in function call. Invalid implicit conversion from contract TransparentUpgradeableProxy to contract ITransparentUpgradeableProxy requested.
>    --> lib/eigenlayer-middleware/test/utils/MockAVSDeployer.sol:310:17:
>     |
> 310 |                 TransparentUpgradeableProxy(payable(address(registryCoordinator))),
>     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> Error: Compilation failed
> ```
> 
> getitng this while compiling my contracts in IS with this as submodule

I wonder if your remappings for openzeppelin are pointing to a different version than in this repo

stevennevins committed:
fix: use libraries with only internal vis (#324)

* fix: revert making library function vis external

* fix: signature checker internal

stevennevins committed:
feat: avs registrar registration flow changes (#318)

* feat: remove both option

* feat: total delegated stake and total slashable stake per quorum config

* test: resolve some breaking changes to tests

* chore: bump core dependency

* chore: bump dependency

* chore: bump to latest slashing mags

* fix: creation of registry coordinator

* test: wip

* feat: integrate registrar interfaces

* test: add function to delegation mock to set operator status

* test: additional test case

* chore: bumping core dep and adding UAM (#325)

* test: use permission controller mock

* chore: label fuzz tests

* test: various fixes from config changes

* chore: remove comment

* test: fix permission controlled functions

* test: fix config issue in integration tests

* test: fix avs directory initialize

* feat: wip prevent m2 registration flows after migration

stevennevins committed:
feat: registration changes part 2

* chore: add note

* fix: remove handling of forceDeregistration

* fix: fix total delegated stake usage

* fix: integration tests

* test: fix remaining integration tests

* test: add back log check

* test: add additional tests for transition to operator sets

* test: add more test cases

* feat: record m2 quorums on migration

* chore: add note about churn support

* fix: prevent operator set registration changes for m2 quorums

* feat: require strings

* chore: add dev note and add require string

stevennevins committed:
chore: bump dependency for slashing mags updates (#329)

nadir-akhtar committed:
Merge branch 'mainnet' into feat/slashing-release-branch

nadir-akhtar committed:
fix: internal slashing security review (#332)

* docs: match file and library name, and add docstrings

* docs: update error strings to match function

* style: forge fmt + moving variables around + fixing error strings

* fix: enforce max quorum count for EjectionManager

* style: more error string fixes + formatting

* fix: correctly set StakeRegistryStorage gap

* style: more error string corrections + typos

stevennevins committed:
fix: revert cases

stevennevins committed:
chore: bump core dependency to pull in latest updates

stevennevins committed:
fix: slashing review fixes (#333)

* fix: withdrawal delay check

* fix: add operator set strategies (#334)

* fix: remove registerOperatorToOperatorSet interface function

* fix: update interface functions with alm interface

* fix: operator set strategies in ALM

* chore: remove todo

* fix: add strategies by stake registry

* fix: params for deregister

* chore: remove old migration functions

* chore: check quorum exists before setting params

* fix: deregister flow for operator set quorums

* test: fix test for DM withdrawal delay blocks

* chore: update all pragmas to 0.8.27 (#336)

* feat: custom require errors for registry coordinator (#337)

* feat: custom require errors for registry coordinator

* ci: update the ci to use the correct compiler settings

stevennevins committed:
chore: bump core dependency

stevennevins committed:
chore: bump forge-std

stevennevins committed:
docs: update readme with note on slashing (#341)

8sunyuan commented on src/RegistryCoordinatorStorage.sol:
```diff
@@ -69,19 +72,24 @@ abstract contract RegistryCoordinatorStorage is IRegistryCoordinator {
     /// @notice the delay in seconds before an operator can reregister after being ejected
     uint256 public ejectionCooldown;
 
+    bool public isOperatorSetAVS;
```
natspec and spacing for consistency

stevennevins replied:
all set

8sunyuan commented on src/RegistryCoordinator.sol:
```diff
@@ -236,22 +256,105 @@ contract RegistryCoordinator is
+        address operator,
+        uint32[] memory operatorSetIds,
+        bytes memory data
+    ) external override onlyWhenNotPaused(PAUSED_REGISTER_OPERATOR) {
```
This needs an `onlyAllocationManager` modifier

stevennevins replied:
yeah can make it a modifier, the check is in the function body

8sunyuan replied:
Ahh, can we either make a modifier or use `_checkAllocationManager` because that internal function is unused.

stevennevins replied:
yeah good call. Forgot i had the internal function

8sunyuan commented on src/RegistryCoordinator.sol:
```diff
@@ -236,22 +256,105 @@ contract RegistryCoordinator is
+    function deregisterOperator(
+        address operator,
+        uint32[] memory operatorSetIds
+    ) external override onlyWhenNotPaused(PAUSED_REGISTER_OPERATOR) {
```
Same `onlyAllocationManager` here

stevennevins replied:
yeah can make it a modifier, the check is in the function body

stevennevins replied:
all set

8sunyuan commented on src/RegistryCoordinator.sol:
```diff
@@ -236,22 +256,105 @@ contract RegistryCoordinator is
+    ) external override onlyWhenNotPaused(PAUSED_REGISTER_OPERATOR) {
+        require(isUsingOperatorSets(), OperatorSetsNotEnabled());
+        for (uint256 i = 0; i < operatorSetIds.length; i++) {
+            require(!isM2Quorum[uint8(operatorSetIds[i])], OperatorSetsNotSupported());
```
can't we just make isM2Quorum a `uint256` and do bitwise and to check this?

stevennevins replied:
don't see a huge benefit of this

8sunyuan commented on src/RegistryCoordinator.sol:
```diff
@@ -443,24 +561,25 @@ contract RegistryCoordinator is
+     *                         INTERNAL FUNCTIONS
+     *
+     */
     struct RegisterResults {
```
can we move this to IRegistryCoordinator

stevennevins replied:
i think this is hear because it's only used in the internal function

8sunyuan commented on src/interfaces/IRegistryCoordinator.sol:
This interface really should contain the register/registerWithChurn/deregister interfaces.

stevennevins replied:
yeah makes sense

8sunyuan commented on src/RegistryCoordinator.sol:
```diff
@@ -57,10 +64,13 @@ contract RegistryCoordinator is
+        IAVSDirectory _avsDirectory,
+        IPauserRegistry _pauserRegistry
+    )
+        RegistryCoordinatorStorage(_serviceManager, _stakeRegistry, _blsApkRegistry, _indexRegistry, _avsDirectory)
```
`AVSDirectory` can be removed as an immutable, its never called

stevennevins replied:
All set

stevennevins committed:
chore: bump core fix iface changes (#352)

* chore: bump dep

* fix: iface changes and param names

0xClandestine committed:
feat: ci storage reports (#347)

* feat: storage reports

* fix: storage report

* fix: storage report

* fix: ci

* fix: ci

* fix: ci

* fix: ci

* fix: ci

stevennevins committed:
fix: prevent calling enable operator sets twice and use internal function for checkALM

stevennevins committed:
chore: address review comments

8sunyuan committed:
refactor: custom errors in middleware (#355)

* refactor: custom errors

* fix: unit tests

8sunyuan commented on src/StakeRegistry.sol:
```diff
@@ -203,20 +228,38 @@ contract StakeRegistry is StakeRegistryStorage {
+     * @param _stakeType The type of stake to track (TOTAL_DELEGATED, TOTAL_SLASHABLE, or BOTH)
+     */
+    function setStakeType(uint8 quorumNumber, StakeType _stakeType) external onlyCoordinatorOwner quorumExists(quorumNumber) {
+        _setStakeType(quorumNumber, _stakeType);
```
Noting that this could potentially break lookups because if there is no stake update according to slashable stake then you may do a delegated stake lookup for the operator.
The BLSSignatureChecker uses `getStakeAtBlockNumberAndIndex`

stevennevins replied:
sg yeah lets remove

stevennevins committed:
chore: review remove setStakeType

stevennevins committed:
fix: wire up stake registry to call add and rm strategies for op set

8sunyuan commented on src/RegistryCoordinator.sol:
```diff
@@ -236,22 +255,106 @@ contract RegistryCoordinator is
+    }
+
+    function isUsingOperatorSets() public view returns (bool) {
+        return isOperatorSetAVS;
```
I think this should just be default to true in the `initialize` function. OperatorSet quorums should be the new normal as the AVSD is planned to be deprecated in some future release.

8sunyuan commented on src/RegistryCoordinator.sol:
```diff
@@ -471,43 +590,104 @@ contract RegistryCoordinator is
+        emit OperatorSocketUpdate(operatorId, socket);
+
+        // If the operator wasn't registered for any quorums, update their status
+        // and register them with this AVS in EigenLayer core (DelegationManager)
```
comment needs updating

ypatil12 commented on src/RegistryCoordinator.sol:
```diff
@@ -172,15 +187,19 @@ contract RegistryCoordinator is
     ) external onlyWhenNotPaused(PAUSED_REGISTER_OPERATOR) {
-        require(operatorKickParams.length == quorumNumbers.length, "RegistryCoordinator.registerOperatorWithChurn: input length mismatch");
-        
+        require(!isUsingOperatorSets(), OperatorSetsEnabled());
```
If we want to enable the same functionality pre and post operatorSets, how do we handle churn once an AVS has started using operatorSets?

8sunyuan commented on src/slashers/VetoableSlasher.sol:
```diff
@@ -0,0 +1,74 @@
+        _cancelSlashingRequest(requestId);
+    }
+
+    function fulfillSlashingRequest(uint256 requestId) external virtual onlySlasher {
```
This modifier is a bit confusing because from the SM perspective this contract is the Slasher but the VetoableSlasher also has a designated slasher address? I guess this can be overrided to be permissionless or to be some other permissioned contract 'slasher'

8sunyuan committed:
refactor: slashing UAM (#357)

* refactor: uam

* feat: add uam interfaces

* fix: lookahead period to blocks

* fix: tests

* chore: storage report

0xClandestine committed:
feat: ci forge coverage (#349)

* feat: ci forge coverage

* fix: ci

0xClandestine committed:
feat: add forge fmt to ci (#363)

* feat: ci forge fmt

* chore: forge fmt

stevennevins committed:
feat: add register with churn type (#360)

* feat: add register with churn type

* fix: tests for encoded registration type

* feat: add ejector support for operator sets

* refactor: handle operator set dereg

* fix: dereg on churn from ALM if needed

8sunyuan committed:
refactor: slashing registry coordinator (#361)

* refactor: slashing registry coordinator

refactor: registry coordinator

refactor: remove sm calls

fix: tests wip

* fix: integration tests

* refactor: internal functions

* fix: remove unused block

* fix: comment out test for now

* build: fmt check run only on src contracts

* fix: call SM only based on m2 quorums (#369)

* feat: update enableOperatorSets flow (#367)

* feat: update enableOperatorSets flow

* fix: dereg op check

* chore: renamed m2 quorums

* fix: msg.sender -> operator

* chore: forge fmt

* fix: tests

---------

Co-authored-by: Yash Patil <40046473+ypatil12@users.noreply.github.com>

8sunyuan committed:
fix: commented out view (#373)

0xClandestine committed:
refactor: natspec + interface changes (#364)

* refactor: natspec + interfaces

refactor: natspec + interfaces

refactor: natspec + interfaces

docs: `RegistryCoordinator` natspec

chore: forge fmt

refactor: named mapping params

docs: natspec `BLSApkRegistry`

docs: natspec `BLSApkRegistry`

docs: natspec `BLSApkRegistry`

refactor: interface structure

refactor: separate `BLSSignatureChecker` storage

refactor: rename -> `IECDSAStakeRegistry`

docs: natspec

docs: natspec `ECDSAStakeRegistry`

docs: natspec `EjectionManager` + separate storage

docs: natspec

docs: natspec `IndexRegistry`

refactor: remove `IRegistry`

chore: forge fmt

chore: make storage-report

docs: natspec

refactor: remove `ISocketUpdater`

refactor: use

refactor: remove todo

refactor: `///` -> `/**`

refactor: improve comment

refactor: rename var

refactor: improve comment

* fix: compiling - tests failing

* refactor: test passing **mostly**

* refactor: add missing gap

* refactor: natspec

* fix: test - state changes moved out of internal

* chore: forge fmt

* refactor: reorganize errors

* refactor: remove override

* refactor: note avsd

* Yash/natspec - address comments (#372)

* feat: natpsec with inheritance

* feat: storage

* chore: format

* chore: fmt

* chore: format

* fix: test

* fix: test

---------

Co-authored-by: Michael Sun <michaelsun97@gmail.com>
Co-authored-by: Yash Patil <40046473+ypatil12@users.noreply.github.com>

stevennevins committed:
chore: resolve merge conflicts

stevennevins committed:
chore: forge fmt

stevennevins committed:
feat: storage gap on socket registry

ypatil12 committed:
chore: remove unused storage; update bindings

stevennevins committed:
chore: nuke storage report script for now on ci

ypatil12 reviewed (approved):
LGTM

gpsanant commented on src/OperatorStateRetriever.sol:
```diff
@@ -22,121 +22,136 @@ contract OperatorStateRetriever {
             uint8 quorumNumber = uint8(quorumNumbers[i]);
-            bytes32[] memory operatorIds = indexRegistry.getOperatorListAtBlockNumber(quorumNumber, blockNumber);
+            bytes32[] memory operatorIds =
+                indexRegistry.getOperatorListAtBlockNumber(quorumNumber, blockNumber);
```
use the ALM's function for operatorSets

gpsanant commented on src/StakeRegistry.sol:
```diff
@@ -710,18 +769,41 @@ contract StakeRegistry is StakeRegistryStorage {
+     * @param quorumNumber The quorum number to set the look ahead period for
+     * @param _lookAheadBlocks The number of blocks to look ahead when checking shares
+     */
+    function _setLookAheadPeriod(uint8 quorumNumber, uint32 _lookAheadBlocks) internal {
```
doesn't check that the quorum is for slashable stake

gpsanant commented on src/StakeRegistry.sol:
```diff
@@ -488,79 +570,83 @@ contract StakeRegistry is StakeRegistryStorage {
      */
+
+    /// @inheritdoc IStakeRegistry
+    function isOperatorSetQuorum(
```
seems like this should be an external function on the registryCoordinator

gpsanant commented on src/StakeRegistry.sol:
```diff
@@ -466,19 +528,39 @@ contract StakeRegistry is StakeRegistryStorage {
+
+        if (stakeTypePerQuorum[quorumNumber] == IStakeRegistryTypes.StakeType.TOTAL_SLASHABLE) {
+            strategyShares = _getSlashableStakePerStrategy(quorumNumber, operator);
+            for (uint256 i = 0; i < stratsLength; i++) {
```
move this loop outside the conditional, it is duplicated

gpsanant commented on src/StakeRegistry.sol:
```diff
@@ -450,14 +492,34 @@ contract StakeRegistry is StakeRegistryStorage {
+    ) internal view returns (uint256[] memory) {
+        address[] memory operators = new address[](1);
+        operators[0] = operator;
+        uint32 beforeTimestamp =
```
beforeBlock
